### PR TITLE
ovirt_nics.py: support adding nic to template

### PR DESCRIPTION
**Pull request by tareqalayan**
before this change adding nic was allowed only to a vm. Now it is
possible to add it to template.

example:
- name: test add nic to template
  ovirt_nics:
    auth: "{{ ovirt_auth }}"
    state: present
    template: golden_mixed_virtio_template
    name: nic1
    interface: virtio
    profile: ovirtmgmt

##### ISSUE TYPE

 - Feature Pull Request
 - New Module Pull Request
 - Bugfix Pull Request
 - Docs Pull Request

##### ANSIBLE VERSION

```

```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by tareqalayan** [comment]
hi Ansible guys, 

2017-04-26 15:30:13 ERROR: test/sanity/pep8/legacy-files.txt:328:1: A201 Remove "lib/ansible/modules/cloud/ovirt/ovirt_nics.py" since it passes the current rule set (75%)

looks to me like some CI issue, can you please take a look?

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py [hunk 1]
```diff
@@ -40,7 +40,12 @@
     vm:
         description:
             - "Name of the Virtual Machine to manage."
-        required: true
+            - "Note that if you provide vm you cannot provide template (vm OR template is required)"
+    template:
+        description:
+            - "Name of the Virtual Machine to manage."
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/ovirt/ovirt_nics.py]: s/Virtual Mahchine/template

**Comment by tareqalayan** [comment]
@mattclay  can you look at this CI error: 
ERROR: Found 1 pep8 issue(s) which need to be resolved:
2017-04-26 17:10:09 ERROR: test/sanity/pep8/legacy-files.txt:328:1: A201 Remove "lib/ansible/modules/cloud/ovirt/ovirt_nics.py" since it passes the current rule set (75%)

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py [hunk 2]
```diff
@@ -40,7 +40,12 @@
     vm:
         description:
             - "Name of the Virtual Machine to manage."
-        required: true
+            - "Note that if you provide vm you cannot provide template (vm OR template is required)"
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/ovirt/ovirt_nics.py]: Please use proper Ansible doc:
```
You must provide either C(vm) parameter or C(template) parameter."
```

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py [hunk 3]
```diff
@@ -40,7 +40,12 @@
...
+    template:
+        description:
+            - "Name of the eemplate to manage."
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/ovirt/ovirt_nics.py]: s/eemplate/template

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py [hunk 4]
```diff
@@ -40,7 +40,12 @@
...
+            - "Note that if you provide template you cannot provide vm (vm OR template is required)"
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/ovirt/ovirt_nics.py]: Same as above.

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py [hunk 5]
```diff
@@ -170,7 +188,8 @@ def main():
             choices=['present', 'absent', 'plugged', 'unplugged'],
             default='present'
         ),
-        vm=dict(required=True),
+        vm=dict(required=False),
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/ovirt/ovirt_nics.py:197]: `required=False` is default, no need to specify.
Please use `required_if`: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/ovirt/ovirt_auth.py#L198

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py [hunk 6]
```diff
@@ -170,7 +188,8 @@ def main():
...
+        template=dict(required=False),
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/ovirt/ovirt_nics.py:198]: same as above

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py [hunk 7]
```diff
@@ -208,13 +240,13 @@ def main():
...
         if profile and module.params['network']:
-            cluster_name = get_link_name(connection, vm.cluster)
+            cluster_name = get_link_name(connection, cluster_name)
             dcs_service = connection.system_service().data_centers_service()
             dc = dcs_service.list(search='Clusters.name=%s' % cluster_name)[0]
             networks_service = dcs_service.service(dc.id).networks_service()
             network = next(
                 (n for n in networks_service.list()
-                if n.name == module.params['network']),
+                 if n.name == module.params['network']),
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/ovirt/ovirt_nics.py]: Why is this needed?

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py [hunk 8]
```diff
@@ -188,18 +207,31 @@ def main():
...
+            template_name = module.params.get('template')
+            template = search_by_name(templates_service, template_name)
+            if template is None:
+                raise Exception("Template '%s' was not found." % template_name)
+
+            service = templates_service.template_service(template.id)
+            cluster_name = template.cluster
+
+        nics_service = service.nics_service()
+        entitynics_module = EntityNicsModule(
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/ovirt/ovirt_nics.py]: Hmm I think this can be simpler, something like this:

```python
if module.params.get('template'):
    entity_name = module.params.get('template')
    collection_service = connection.system_service().templates_service()
elif module.params.get('vm'):
    entity_name = module.params.get('vm')
    collection_service = connection.system_service().vms_service()
            
# TODO: We have to modify the search_by_name function to accept raise_error=True/False,
entity = search_by_name(collection_service, entity_name)
if entity is None:
    raise Exception("Vm/Template '%s' was not found." % entity_name)

service = collection_service.service(entity.id)
cluster_id = entity.cluster
```

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py [hunk 9]
```diff
@@ -208,13 +240,13 @@ def main():
         # Find vNIC id of the network interface (if any):
         profile = module.params.get('profile')
         if profile and module.params['network']:
-            cluster_name = get_link_name(connection, vm.cluster)
+            cluster_name = get_link_name(connection, cluster_name)
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/ovirt/ovirt_nics.py]: `cluster_name = get_link_name(connection, cluster_name)`
->
`cluster_name = get_link_name(connection, cluster_id)`

**Review by machacekondra — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py (hunk 2 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_nics.py]: done

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py (hunk 1 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_nics.py]: done

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py (hunk 3 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_nics.py]: done

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py (hunk 4 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_nics.py]: done

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py (hunk 5 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_nics.py:197]: nice.
done

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py (hunk 6 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_nics.py:198]: done

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py (hunk 9 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_nics.py]: done

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py (hunk 7 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_nics.py]: mistake

File: lib/ansible/modules/cloud/ovirt/ovirt_nics.py [hunk 10]
```diff
@@ -170,13 +188,17 @@ def main():
...
-        vm=dict(required=True),
+        vm=dict(required=False),
+        template=dict(required=False),
         name=dict(required=True),
         interface=dict(default=None),
         profile=dict(default=None),
         network=dict(default=None),
         mac_address=dict(default=None),
     )
+    required_of=[
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/ovirt/ovirt_nics.py]: Please verify, before push. `required_of` -> `required_if`

Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_nics.py]: fixed. it is required_one_of 
thanks,

**Review by machacekondra — APPROVED** [review]

**Comment by machacekondra** [comment]
shipit

**Review by myakove — APPROVED** [review]

**Comment by machacekondra** [comment]
@mwperina @ryansb

**Review by mwperina — APPROVED** [review]
👍
